File: python/LongestIncreasingSubsequence.py
#https://leetcode.com/problems/longest-increasing-subsequence/
class SolutionLIS:
    """
    @param nums: The integer array
    @return: The length of LIS (longest increasing subsequence)
    """
    def longestIncreasingSubsequence_DP(self, nums):
        memo, ans = [1] * len(nums), 1
        for i in range(1, len(nums)):
            for j in range(0, i):
                if nums[j] < nums[i]:
                    memo[i] = max(memo[i], memo[j] + 1)
            ans = max(ans, memo[i])
        return ans if len(nums) > 0 else 0
        
    def longestIncreasingSubsequence_NlogN(self, nums):
        ans, tmp = 0, 0
        for i in range(1, len(nums)):
            #Find rightmost value in sublist of nums that is no less than nums[i]
            tmp = bisect.bisect_left(nums[0 : ans + 1], nums[i])
            if tmp > ans:
                ans += 1
                nums[ans] = nums[i]
            else:
                nums[tmp] = nums[i]
        return ans + 1 if len(nums) > 0 else 0

File: java/LongestIncreasingSubsequence.java
import java.util.Arrays;

//https://leetcode.com/problems/longest-increasing-subsequence
class SolutionLongestIncreasingSubsequence {
    //NLogN BS
    public int lengthOfLIS(int[] nums) {
        int len = nums.length, ans = 0, r = 0, l = 0, sz = 0, m, t;
        int[] seq = new int[len];
        for(int i : nums) {
            if(0 == sz) seq[sz++] = i;
            else {
                l = 0;
                r = sz - 1;
                t = sz;
                while(l <= r) {
                    m = l + (r - l) / 2;
                    if(seq[m] >= i) {
                        t = m;
                        r = m - 1;
                    }
                    else
                        l = m + 1;
                }
                seq[t] = i;
                if(sz == t) ++sz;
            }
        }
        return 1 > len ? 0 : sz;
    }
    //O(N^2) DP
    public int lengthOfLIS1(int[] nums) {
        int len = nums.length, ans = 0;
        int[] memo = new int[len];
        Arrays.fill(memo, 1);
        for(int i = 1; i < len; ++i) {
            for(int j = i - 1; j > -1; --j)
                if(nums[j] < nums[i]) memo[i] = Math.max(memo[i], memo[j] + 1);
            ans = Math.max(ans, memo[i]);
        }
        return 1 == len ? 1 : ans;
    }
}
